transport/src/main/java/io/camunda/zeebe/transport/stream/impl/ClientStreamManager.java

```diff
@@ -53,20 +53,22 @@ UUID add(
       final M metadata,
       final ClientStreamConsumer clientStreamConsumer) {
     final var streamId = UUID.randomUUID();
-    final var clientStreamMeta =
-        new ClientStream<>(streamId, streamType, metadata, clientStreamConsumer);
 
     // add first in memory to handle case of new broker while we're adding
-    registry.add(clientStreamMeta);
-    requestManager.openStream(clientStreamMeta, servers);
+    final AggregatedClientStream<M> serverStream =
+        registry.addClient(streamId, streamType, metadata, clientStreamConsumer);
+    LOG.debug("Added client stream [{}] to stream [{}]", streamId, serverStream.getStreamId());
+    requestManager.openStream(serverStream, servers);
```

Comment:
:wrench: I think this may lead to more requests than necessary being sent out. If I add two equivalent streams one after the other, and the first one had not finished connecting, it will send out requests again, since we don't have a "connecting" state. One option is to add that (so we can already tell if we're trying to connect), but even that means adding a little bit of useless "work" to the actor/executor. So one thing we could do is return also an `Optional` like on the remove case, and only return something if it's the "first" time. That may introduce more edge cases though (e.g. what if an error occurred then, and we didn't open it)...so maybe the "connecting" thing is more interesting. At any rate, we can add that in a second pass as well if it's too much work, but we should extract it as a follow up in that case.

Reply:
I think returning something only if it's the first time, would be good enough here. I didn't want to do it before because it is a bit confusing, but ended up using same approach for remove. So it should be fine. I will update this PR, as it is not a big change.

> That may introduce more edge cases though (e.g. what if an error occurred then, and we didn't open it)

We don't consider this case now. We retry indefinitely until it succeeds. So I guess it is ok to ignore that case for now. We should handle errors in a different way anyway, like for eg:- close all client streams if it cannot connect to servers.

Reply:
I ended up doing something different, along the lines of adding a new state. https://github.com/camunda/zeebe/pull/12402/commits/e17106ba2f34c7d2ae8308844f16031a96452566